pkg/webhook/bd_replacement.go

```diff
@@ -0,0 +1,383 @@
+	for _, oldRg := range oldPoolSpec.RaidGroups {
+		for _, newRg := range newPoolSpec.RaidGroups {
+			if IsRaidGroupCommon(oldRg, newRg) {
+				if IsBlockDeviceReplacementCase(oldRg, newRg) {
+					if ok, msg := bdr.IsBDReplacementValid(newRg, oldRg); !ok {
+						return false, msg
+					} else {
+						newBD := GetNewBDFromRaidGroups(newRg, oldRg)
+						for k, v := range newBD {
+							newToOldBd[k] = v
+						}
+					}
+				}
+			}
+		}
+	}
+
+	for newBD, oldBD := range newToOldBd {
+		err := createBDC(newBD, oldBD, bdr.OldCSPC)
+		if err != nil {
+			return false, err.Error()
+		}
+	}
+	return true, ""
+}
+
+func IsRaidGroupCommon(rgOld, rgNew apis.RaidGroup) bool {
+	oldBdMap := make(map[string]bool)
+	for _, bd := range rgOld.BlockDevices {
+		oldBdMap[bd.BlockDeviceName] = true
+	}
+
+	for _, bd := range rgNew.BlockDevices {
+		if oldBdMap[bd.BlockDeviceName] {
+			return true
+		}
+	}
+	return false
+}
+
+// IsBlockDeviceReplacementCase returns true if the edit/update of CSPC can trigger a blockdevice
+// replacement.
+func IsBlockDeviceReplacementCase(newRaidGroup, oldRaidGroup apis.RaidGroup) bool {
+	count := GetNumberOfDiskReplaced(newRaidGroup, oldRaidGroup)
+	if count >= 1 {
+		return true
+	}
+	return false
+}
+
+// IsMoreThanOneDiskReplaced returns true if more than one disk is replaced in the same raid group.
+func GetNumberOfDiskReplaced(newRG apis.RaidGroup, oldRG apis.RaidGroup) int {
+	var count int
+	oldBlockDevicesMap := make(map[string]bool)
+	for _, bdOld := range oldRG.BlockDevices {
+		oldBlockDevicesMap[bdOld.BlockDeviceName] = true
+	}
+	for _, newBD := range newRG.BlockDevices {
+		if !oldBlockDevicesMap[newBD.BlockDeviceName] {
+			count++
+		}
+	}
+	return count
+}
+
+// IsBDReplacementValid validates for BD replacement.
+func (bdr *BlockDeviceReplacement) IsBDReplacementValid(newRG apis.RaidGroup, oldRG apis.RaidGroup) (bool, string) {
+
+	// Not more than 1 bd should be replaced in a raid group.
+	if IsMoreThanOneDiskReplaced(newRG, oldRG) {
+		return false, "cannot replace more than one blockdevice in a raid group"
+	}
+
+	// The incoming BD for replacement should not pe present in the current CSPC.
+	if bdr.IsNewBDPresentOnCurrentCSPC(newRG, oldRG) {
+		return false, "the new blockdevice intended to use for replacement is already a part of the current cspc"
+	}
+
+	// No background replacement should be going on in the raid group undergoing replacement.
+	if ok, err := bdr.IsExistingReplacmentInProgress(oldRG); ok {
+		return false, fmt.Sprintf("cannot replace blockdevice as a "+
+			"background replacement may be in progress in the raid group: %s", err.Error())
+	}
+
+	// The incoming BD should be a valid entry if
+	// 1. The BD does not have a BDC.
+	// 2. The BD has a BDC with the current CSPC label and there is no successor of this BD
+	//    present in the CSPC.
+	if !bdr.AreNewBDsValid(newRG, oldRG, bdr.OldCSPC) {
+		return false, "the new blockdevice intended to use for replacement in invalid"
```

Comment:
I think here error needs to be updated with exact error caused validation.

Reply:
Will fix in next PR immediately to address debuggability aspects